dict: dict_reset() delete all elements using iteration

tshacked opened the pull request:
Enhance dict_reset() imp by deleting all elements using iteration

Fixes: #1536
Change-Id: Ib4d4f80bd30d52c891eb0fd4b563db19134e2328
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked committed:
dict: dict_reset() delete all elements using iteration

Enhance dict_reset() imp by deleting all elements using iteration

Fixes: #1536
Change-Id: Ib4d4f80bd30d52c891eb0fd4b563db19134e2328
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

mykaul commented on libglusterfs/src/dict.c:
```diff
@@ -1468,7 +1468,31 @@ dict_reset(dict_t *dict)
+    data_pair_t *curr = dict->members_list;
+    data_pair_t *next = NULL;
+
+    while (curr != NULL) {
```
The code is essentially like dict_destroy(), just without the deletion of the dictionary, no? I wonder if we can reuse some code. Otherwise, looks OK.

tshacked replied:
correct, i'll place the iteration block in a separate function for reusing.

tshacked committed:
dict: add dict_clear_data() to resue in dict_reset/destroy()

add dict_clear_data() to resue in dict_reset/destroy()
change dict_destroy to static

Fixes: #1536
Change-Id: I015d16f7abc9b03624cafbb838c7efb29a5eaede
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

aspandey commented on libglusterfs/src/dict.c:
```diff
@@ -1469,28 +1476,14 @@ dict_reset(dict_t *dict)
 
-    for (int i = 0; i < dict->hash_size; i++)
+    dict_clear_data(dict);
+    for (; i < dict->hash_size; i++)
```
What performance are we gaining by not mentioning i=0 in this for loop? 
I think we should have written it here.

tshacked replied:
On my 1st commit, I declared `i` inside the loop and got the `error: 'for' loop initial declarations are only allowed in C99 mode` compilation error in smoke-centos7 job - https://build.gluster.org/job/gh_smoke-centos7/766/console.
I mv both declaration+initialization outside the loop just for fixing the error.
will move the initialization inside the loop if this is the convention..

aspandey commented on libglusterfs/src/dict.c:
```diff
@@ -684,7 +684,28 @@ dict_deln(dict_t *this, char *key, const int keylen)
-void
+/* removes and free all data_pair_t elements.
+ * has to be called with this->lock held */
+static void
```
Here is the point. this->count gives you the number of pairs in this dictionary.
So while in a while loop we are destroying/freeing data pairs, we should also keep decreasing it.
Or at the end make it 0.

tshacked replied:
Agree, will add this to the function.

aspandey commented on libglusterfs/src/dict.c:
```diff
@@ -693,26 +714,12 @@ dict_destroy(dict_t *this)
-        } else {
-            this->free_pair.key = NULL;
-        }
-        total_pairs++;
```
We should get this number from this->count before calling dict_clear_data and then use it. Specially if you are going to implement my previous comment.

tshacked replied:
- agree about `this->count`
- the usage of `total_pairs` (and entire block) comes from the original imp of `dict_destroy`

tshacked committed:
dict: dict_reset() code review fixes

- reset this->count at dict_clear_data()

Change-Id: Ifef6c9d7ca4e447c7314d96463f0ea6ab154f22f
Signed-off-by: Tamar Shacked <tshacked@redhat.com>

tshacked commented:
/run regression
